File: pools/allocating_strategies/strategies/src/ipv4_prefix_strategy.js
import {
    hostsInMask,
    inet_aton,
    inet_ntoa,
    prefixToStr,
    subnetAddresses
} from "./ipv4utils";

// framework managed constants
var currentResources = []
var resourcePoolProperties = {}
var userInput = {}
// framework managed constants

// STRATEGY_START

/*
IPv4 prefix allocation strategy

- Expects IPv4 prefix resource type to have 2 properties of type int ["address:string", "mask:int"]
- userInput.desiredSize is a required parameter e.g. desiredSize == 10  ---produces-prefix-of--->  192.168.1.0/28
- userInput.subnet is an optional parameter specifying whether allocated prefix will be used as a real subnet or just
  as an IP pool. Essentially whether to count subnet address and broadcast into the range or not. If set to true, the resulting
  prefix will have a capacity of desiredSize + 2. Defaults to false
- Logs utilisation stats
- Allocates previously freed prefixes
- All addresses from parent prefix are used, including the first and last one
 */

// compare prefixes based on their broadcast address
function comparePrefix(prefix1, prefix2) {
    let endOfP1 = inet_aton(prefix1.address) + subnetAddresses(prefix1.prefix)
    let endOfP2 = inet_aton(prefix2.address) + subnetAddresses(prefix2.prefix)
    return endOfP1 - endOfP2
}

// sum up capacity of an array of addresses
function prefixesCapacity(currentResources) {
    let width = 0
    for (let allocatedPrefix of currentResources) {
        width += subnetAddresses(allocatedPrefix.prefix)
    }
    return width
}

// calculate utilized capacity based on previously allocated prefixes + a newly allocated prefix
function utilizedCapacity(allocatedRanges, newlyAllocatedRangeCapacity) {
    return prefixesCapacity(allocatedRanges) + newlyAllocatedRangeCapacity
}

// calculate free capacity based on previously allocated prefixes
function freeCapacity(parentPrefix, utilisedCapacity) {
    return subnetAddresses(parentPrefix.prefix) - utilisedCapacity
}

function capacity() {
    let totalCapacity = hostsInMask(resourcePoolProperties.address, resourcePoolProperties.prefix);
    let allocatedCapacity = 0;
    let resource;
    let subnetItself = userInput.subnet ? 1 : 0;

    for (resource of currentResources) {
        allocatedCapacity += hostsInMask(resource.Properties.address, resource.Properties.prefix);
    }

    return { freeCapacity: totalCapacity - allocatedCapacity + subnetItself, utilizedCapacity: allocatedCapacity };
}

// log utilisation stats
function logStats(newlyAllocatedPrefix, parentRange, allocatedPrefixes = [], level = "log") {
    let newlyAllocatedPrefixCapacity = 0
    if (newlyAllocatedPrefix) {
        newlyAllocatedPrefixCapacity = subnetAddresses(newlyAllocatedPrefix.prefix)
    } else {
        newlyAllocatedPrefixCapacity = 0
    }

    let utilisedCapacity = utilizedCapacity(allocatedPrefixes, newlyAllocatedPrefixCapacity)
    let remainingCapacity = freeCapacity(parentRange, utilisedCapacity)
    let utilPercentage
    if (remainingCapacity === 0) {
        utilPercentage = 100.0
    } else {
        utilPercentage = (utilisedCapacity / subnetAddresses(parentRange.prefix)) * 100
    }
    console[level]("Remaining capacity: " + remainingCapacity)
    console[level]("Utilised capacity: " + utilisedCapacity)
    console[level](`Utilisation: ${utilPercentage.toFixed(1)}%`)
}

function prefixesToString(currentResourcesUnwrapped) {
    let prefixesToStr = ""
    for (let allocatedPrefix of currentResourcesUnwrapped) {
        prefixesToStr += prefixToStr(allocatedPrefix)
        prefixesToStr += prefixToRangeStr(allocatedPrefix)
        prefixesToStr += ", "
    }
    return prefixesToStr
}

function prefixToRangeStr(prefix) {
    return `[${prefix.address}-${inet_ntoa(inet_aton(prefix.address) + subnetAddresses(prefix.prefix) - 1)}]`
}

function calculateDesiredSubnetMask() {
    let newSubnetBits = Math.ceil(Math.log(userInput.desiredSize) / Math.log(2))
    let newSubnetMask = 32 - newSubnetBits
    let newSubnetCapacity = subnetAddresses(newSubnetMask)
    return {newSubnetMask, newSubnetCapacity};
}

// calculate the nearest possible address for a subnet where mask === newSubnetMask
//  that is outside of allocatedSubnet
function findNextFreeSubnetAddress(allocatedSubnet, newSubnetMask) {
    // find the first address after currently iterated allocated subnet
    let nextAvailableAddressNum = inet_aton(allocatedSubnet.address) + subnetAddresses(allocatedSubnet.prefix)
    // remove any bites from the address above after newSubnetMask
    let newSubnetMaskNegative = 32 - newSubnetMask
    let possibleSubnetNum = (nextAvailableAddressNum >>> newSubnetMaskNegative) << newSubnetMaskNegative
    // keep going until we find an address outside of currently iterated allocated subnet
    while (nextAvailableAddressNum > possibleSubnetNum) {
        possibleSubnetNum = ((possibleSubnetNum >>> newSubnetMaskNegative) + 1) << newSubnetMaskNegative
    }
    return possibleSubnetNum;
}

// main
function invoke() {
    let rootPrefixParsed = resourcePoolProperties
    if (rootPrefixParsed == null) {
        console.error("Unable to extract root prefix from pool name: " + rootPrefix)
        return null
    }
    let rootAddressStr = rootPrefixParsed.address
    let rootMask = rootPrefixParsed.prefix
    let rootPrefixStr = prefixToStr(rootPrefixParsed)
    let rootCapacity = subnetAddresses(rootMask)
    let rootAddressNum = inet_aton(rootAddressStr)

    if (!userInput.desiredSize) {
        console.error("Unable to allocate subnet from root prefix: " + rootPrefixStr +
            ". Desired size of a new subnet size not provided as userInput.desiredSize")
        return null
    }

    if (userInput.desiredSize < 2) {
        console.error("Unable to allocate subnet from root prefix: " + rootPrefixStr +
            ". Desired size is invalid: " + userInput.desiredSize + ". Use values >= 2")
        return null
    }

    if (userInput.subnet === true) {
        // reserve subnet address and broadcast
        userInput.desiredSize += 2
    }

    // Calculate smallest possible subnet mask to fit desiredSize
    let {newSubnetMask, newSubnetCapacity} = calculateDesiredSubnetMask();

    // unwrap and sort currentResources
    let currentResourcesUnwrapped = currentResources.map(cR => cR.Properties)
    currentResourcesUnwrapped.sort(comparePrefix)

    let possibleSubnetNum = rootAddressNum
    // iterate over allocated subnets and see if a desired new subnet can be squeezed in
    for (let allocatedSubnet of currentResourcesUnwrapped) {

        let allocatedSubnetNum = inet_aton(allocatedSubnet.address)
        let chunkCapacity = allocatedSubnetNum - possibleSubnetNum
        if (chunkCapacity >= userInput.desiredSize) {
            // there is chunk with sufficient capacity between possibleSubnetNum and allocatedSubnet.address
            let newlyAllocatedPrefix = {
                "address": inet_ntoa(possibleSubnetNum),
                "prefix": newSubnetMask
            }
            // FIXME How to pass these stats ?
            // logStats(newlyAllocatedPrefix, rootPrefixParsed, currentResourcesUnwrapped)
            return newlyAllocatedPrefix
        }

        // move possible subnet start to a valid address outside of allocatedSubnet's addresses and continue the search
        possibleSubnetNum = findNextFreeSubnetAddress(allocatedSubnet, newSubnetMask);
    }

    // check if there is any space left at the end of parent range
    if (possibleSubnetNum + newSubnetCapacity <= rootAddressNum + rootCapacity) {
        // there sure is some space, use it !
        let newlyAllocatedPrefix = {
            "address": inet_ntoa(possibleSubnetNum),
            "prefix": newSubnetMask
        }
        // FIXME How to pass these stats ?
        // logStats(newlyAllocatedPrefix, rootPrefixParsed, currentResourcesUnwrapped)
        return newlyAllocatedPrefix
    }

    // no suitable range found
    console.error("Unable to allocate Ipv4 prefix from: " + rootPrefixStr +
        ". Insufficient capacity to allocate a new prefix of size: " + userInput.desiredSize)
    console.error("Currently allocated prefixes: " + prefixesToString(currentResourcesUnwrapped))
    logStats(null, rootPrefixParsed, currentResourcesUnwrapped, "error")
    return null
}

// STRATEGY_END

// For testing purposes
function invokeWithParams(currentResourcesArg, resourcePoolArg, userInputArg) {
    currentResources = currentResourcesArg
    resourcePoolProperties = resourcePoolArg
    userInput = userInputArg
    return invoke()
}

// For testing purposes
function invokeWithParamsCapacity(currentResourcesArg, resourcePoolArg, userInputArg) {
    currentResources = currentResourcesArg
    resourcePoolProperties = resourcePoolArg
    userInput = userInputArg
    return capacity()
}
exports.invoke = invoke
exports.capacity = capacity
exports.invokeWithParams = invokeWithParams
exports.invokeWithParamsCapacity = invokeWithParamsCapacity
exports.utilizedCapacity = utilizedCapacity
exports.freeCapacity = freeCapacity
// For testing purposes

File: pools/allocating_strategies/strategies/src/ipv4_prefix_strategy.go
package src

import (
	"github.com/pkg/errors"
	"github.com/vektah/gqlparser/v2/gqlerror"
	"math"
	"net"
	"reflect"
	"sort"
	"strconv"
)

type Ipv4Struct struct {
	address string
	prefix  int
}

type Ipv4Prefix struct {
	currentResources       []map[string]interface{}
	resourcePoolProperties map[string]interface{}
	userInput              map[string]interface{}
}

func NewIpv4Prefix(currentResources []map[string]interface{},
	resourcePoolProperties map[string]interface{},
	userInput map[string]interface{}) Ipv4Prefix {
	return Ipv4Prefix{currentResources, resourcePoolProperties, userInput}
}

func (ipv4prefix *Ipv4Prefix) UtilizedCapacity(
	allocatedRanges []map[string]interface{},
	newlyAllocatedRangeCapacity int) int {
	return prefixesCapacity(allocatedRanges) + newlyAllocatedRangeCapacity
}

// FreeCapacity calculate free capacity based on previously allocated prefixes
func (ipv4prefix *Ipv4Prefix) FreeCapacity(
	parentPrefix map[string]interface{},
	utilisedCapacity int) int {
	prefix, _ := parentPrefix["prefix"]
	return subnetAddresses(prefix.(int)) - utilisedCapacity
}

func getAddressAndPrefixFromCurrentResource(currentResource map[string]interface{}) (string, int, error) {
	value, ok := currentResource["Properties"]
	if !ok {
		return "", 0, errors.New("Unable to extract properties from resource")
	}
	address, ok := value.(map[string]interface{})["address"]
	if !ok {
		return "", 0, errors.New("Unable to extract address resource from properties")
	}
	prefix, ok := value.(map[string]interface{})["prefix"]
	if !ok {
		return "", 0, errors.New("Unable to extract prefix resource from properties")
	}
	prefix, err := NumberToInt(prefix)
	if err != nil {
		return "", 0, err
	}
	return address.(string), prefix.(int), nil
}

func (ipv4prefix *Ipv4Prefix) Capacity() (map[string]interface{}, error) {
	if ipv4prefix.resourcePoolProperties == nil {
		return nil, errors.New("Unable to extract resources")
	}
	rootAddressStr, ok := ipv4prefix.resourcePoolProperties["address"]
	if !ok {
		return nil, errors.New("Unable to extract address resource")
	}
	rootMask, ok := ipv4prefix.resourcePoolProperties["prefix"]
	if !ok {
		return nil, errors.New("Unable to extract prefix resources")
	}

	isSubnet, isSubnetOk := ipv4prefix.resourcePoolProperties["subnet"].(bool)
	totalCapacity := hostsInMask(rootAddressStr.(string), rootMask.(int))
	allocatedCapacity := 0

	if isSubnetOk && isSubnet && (rootMask.(int) == 31 || rootMask.(int) == 32) {
		var result = make(map[string]interface{})
		result["freeCapacity"] = strconv.Itoa(0)
		result["utilizedCapacity"] = strconv.Itoa(0)

		return result, nil
	}

	if isSubnetOk && isSubnet == false && rootMask.(int) != 31 && rootMask.(int) != 32 {
		// we are adding 2 because total capacity is not only host addresses but also network and broadcast addresses
		// 31 and 32 prefixes return constant value, so there is not needed this calculation
		totalCapacity += 2
	}

	for _, resource := range ipv4prefix.currentResources {
		address, prefix, err := getAddressAndPrefixFromCurrentResource(resource)
		if err != nil {
			return nil, err
		}

		if prefix == 31 || prefix == 32 {
			allocatedCapacity += hostsInMask(address, prefix)
		} else {
			allocatedCapacity += hostsInMask(address, prefix) + 2
		}
	}

	var freeCapacity float64
	// we are handling the edge case when allocated capacity is same as total capacity,
	// then we need to add network and broadcast address (+2) of the main subnet
	if totalCapacity+2 == allocatedCapacity {
		freeCapacity = 0
	} else {
		freeCapacity = float64(totalCapacity - allocatedCapacity)
	}

	var result = make(map[string]interface{})
	result["freeCapacity"] = strconv.FormatFloat(freeCapacity, 'g', 30, 64)
	result["utilizedCapacity"] = strconv.Itoa(allocatedCapacity)

	return result, nil
}

func prefixesCapacity(currentResources []map[string]interface{}) int {
	width := 0
	for _, allocatedPrefix := range currentResources {
		value, _ := allocatedPrefix["prefix"]
		width += subnetAddresses(value.(int))
	}
	return width
}

// calculate the nearest possible address for a subnet where mask === newSubnetMask
//
//	that is outside of allocatedSubnet
func findNextFreeSubnetAddress(allocatedSubnet Ipv4Struct, newSubnetMask int) int {
	// find the first address after currently iterated allocated subnet
	addressNumber, _ := InetAton(allocatedSubnet.address)
	nextAvailableAddressNum := addressNumber + subnetAddresses(allocatedSubnet.prefix)
	// remove any bites from the address above after newSubnetMask
	newSubnetMaskNegative := 32 - newSubnetMask
	possibleSubnetNum := int(uint(nextAvailableAddressNum)>>newSubnetMaskNegative) << newSubnetMaskNegative
	// keep going until we find an address outside of currently iterated allocated subnet
	for nextAvailableAddressNum > int(possibleSubnetNum) {
		possibleSubnetNum = (int(uint(possibleSubnetNum)>>newSubnetMaskNegative) + 1) << newSubnetMaskNegative
	}
	return possibleSubnetNum
}

func calculateDesiredSubnetIpv4Mask(desiredSize int) (int, int) {
	newSubnetBits := math.Ceil(math.Log(float64(desiredSize)) / math.Log(2))
	newSubnetMask := 32 - newSubnetBits
	newSubnetCapacity := subnetAddresses(int(newSubnetMask))
	return int(newSubnetMask), newSubnetCapacity
}

func isIPv4AddrNetwork(addr string, prefix int) (bool, string, error) {
	_, ipNet, ipErr := net.ParseCIDR(fmt.Sprintf("%s/%d", addr, prefix))

	if ipErr != nil {
		return false, ipNet.IP.String(), ipErr
	}

	if addr == ipNet.IP.String() {
		return true, ipNet.IP.String(), nil
	}

	return false, ipNet.IP.String(), nil
}

func nextFreeNetworkAddressAfter(networkAddress string, prefix int, capacity int, allocatedResources []Ipv4Struct) string {
	networkAddressNum, _ := InetAton(networkAddress)
	var possibleSubnetNum = networkAddressNum

	for _, allocatedSubnet := range allocatedResources {
		allocatedSubnetNum, _ := InetAton(allocatedSubnet.address)
		chunkCapacity := allocatedSubnetNum - possibleSubnetNum

		fmt.Println(inetNtoa(possibleSubnetNum))

		if chunkCapacity >= capacity && allocatedSubnetNum > networkAddressNum {
			return inetNtoa(possibleSubnetNum)
		}

		// move possible subnet start to a valid address outside allocatedSubnet's addresses and continue the search
		possibleSubnetNum = findNextFreeSubnetAddress(allocatedSubnet, prefix)
	}

	if networkAddressNum > possibleSubnetNum {
		return networkAddress
	} else {
		_, netAddr, _ := isIPv4AddrNetwork(inetNtoa(possibleSubnetNum), prefix)
		return netAddr
	}
}

func (ipv4prefix *Ipv4Prefix) Invoke() (map[string]interface{}, error) {
	if ipv4prefix.resourcePoolProperties == nil {
		return nil, errors.New("Unable to extract resources")
	}
	rootAddressStr, ok := ipv4prefix.resourcePoolProperties["address"]
	if !ok {
		return nil, errors.New("Unable to extract address resource")
	}
	rootMask, ok := ipv4prefix.resourcePoolProperties["prefix"]
	if !ok {
		return nil, errors.New("Unable to extract prefix resources")
	}
	isSubnet, ok := ipv4prefix.resourcePoolProperties["subnet"]
	if !ok {
		return nil, errors.New("Unable to extract subnet resources")
	}
	rootMask, err := NumberToInt(rootMask)
	if err != nil {
		return nil, err
	}
	rootPrefixStr := prefixToStr(ipv4prefix.resourcePoolProperties)
	rootCapacity := subnetAddresses(rootMask.(int))
	rootAddressNum, err := InetAton(rootAddressStr.(string))
	if err != nil {
		return nil, err
	}

	value, ok := ipv4prefix.userInput["desiredSize"]
	if !ok {
		return nil, errors.New("Unable to allocate subnet from root prefix: " + rootPrefixStr +
			". Desired size of a new subnet size not provided as userInput.desiredSize")
	}

	var desiredSize interface{}
	var desiredSizeErr error
	if isSubnet.(bool) {
		desiredSize, desiredSizeErr = NumberToInt(value)
		if desiredSizeErr == nil {
			desSize, ok := desiredSize.(int)

			if !ok {
				return nil, gqlerror.Errorf("Unable to claim resource: usrInput.desiredSize was sent in bad format. Required format is int or string and was: %s", reflect.TypeOf(desiredSize))
			}

			desiredSize = desSize + 2
		} else {
			return nil, gqlerror.Errorf("Unable to claim resource: %s", desiredSizeErr)
		}
	} else {
		desiredSize, desiredSizeErr = NumberToInt(value)

		if desiredSizeErr != nil {
			return nil, gqlerror.Errorf("Unable to claim resource: %s", desiredSizeErr)
		}
	}
	desiredValue, isDesiredValueOk := ipv4prefix.userInput["desiredValue"]

	if !isDesiredValueOk {
		desiredValue = nil
	}

	if desiredSizeErr != nil || err != nil {
		return nil, err
	}
	if desiredSize.(int) < 0 {
		return nil, errors.New("Unable to allocate subnet from root prefix: " + rootPrefixStr +
			". Desired size is invalid: " + strconv.Itoa(desiredSize.(int)) + ". Use values >= 2")
	}
	newSubnetMask, newSubnetCapacity := calculateDesiredSubnetIpv4Mask(desiredSize.(int))

	if isSubnet.(bool) && (newSubnetMask == 31 || newSubnetMask == 32) {
		return nil, errors.Errorf("It is not possible to allocate resource with prefix %d, together with subnet set as true", newSubnetMask)
	}

	if isSubnet.(bool) && rootCapacity < desiredSize.(int) {
		return nil, errors.New("Unable to allocate Ipv4 prefix from: " + rootPrefixStr + ". " +
			"Insufficient capacity to allocate a new prefix of size: " + strconv.Itoa(desiredSize.(int)-2) + "\n" +
			"Currently allocated addresses: " + addressesToStr(ipv4prefix.currentResources))
	}

	if err != nil && desiredValue != nil {
		return nil, errors.New("We weren't able to handle formatting of provided inputs, that were of incorrect shape")
	}

	var currentResourcesStruct []Ipv4Struct
	for _, resource := range ipv4prefix.currentResources {
		address, prefix, err := getAddressAndPrefixFromCurrentResource(resource)
		if err != nil {
			return nil, err
		}
		currentResourcesStruct = append(currentResourcesStruct, Ipv4Struct{address: address, prefix: prefix})
	}

	// compare prefixes based on their broadcast address
	sort.Slice(currentResourcesStruct, func(i, j int) bool {
		address1Num, _ := InetAton(currentResourcesStruct[i].address)
		address2Num, _ := InetAton(currentResourcesStruct[j].address)
		endOfP1 := address1Num + subnetAddresses(currentResourcesStruct[i].prefix)
		endOfP2 := address2Num + subnetAddresses(currentResourcesStruct[j].prefix)
		return endOfP1 < endOfP2
	})

	possibleSubnetNum := rootAddressNum
	var result = make(map[string]interface{})

	if desiredValue != nil {
		isNetwork, networkAddr, ipErr := isIPv4AddrNetwork(desiredValue.(string), newSubnetMask)

		if ipErr != nil {
			return nil, ipErr
		}

		if !isNetwork {
			nextFreeNetworkAddress := nextFreeNetworkAddressAfter(networkAddr, newSubnetMask, newSubnetCapacity, currentResourcesStruct)
			return nil, errors.Errorf("You provided invalid network address. Network address should be %s", nextFreeNetworkAddress)
		}

		if len(currentResourcesStruct) > 0 {
			desiredValueNum, er := InetAton(desiredValue.(string))
			lastResource := currentResourcesStruct[len(currentResourcesStruct)-1]
			lastResourceSubnetNum, e := InetAton(lastResource.address)

			if er != nil || e != nil {
				return nil, errors.New("We weren't able to handle formatting of provided inputs, that were of incorrect format")
			}

			if lastResourceBroadcastNum := subnetLastAddress(lastResourceSubnetNum, lastResource.prefix); desiredValueNum > lastResourceBroadcastNum {
				if desiredValueNum+newSubnetCapacity <= rootAddressNum+rootCapacity {
					// there is chunk with sufficient capacity between possibleSubnetNum and allocatedSubnet.address
					result["address"] = desiredValue
					result["prefix"] = newSubnetMask
					result["subnet"] = isSubnet

					return result, nil
				}
			}
		} else {
			desiredValueNum, e := InetAton(desiredValue.(string))

			if e != nil {
				return nil, errors.New("We weren't able to handle formatting of provided inputs, that were of incorrect format")
			}

			if desiredValueNum+newSubnetCapacity <= rootAddressNum+rootCapacity {
				// there is chunk with sufficient capacity between possibleSubnetNum and allocatedSubnet.address
				result["address"] = desiredValue
				result["prefix"] = newSubnetMask
				result["subnet"] = isSubnet

				return result, nil
			}

			return nil, errors.New("Unable to allocate Ipv4 prefix from: " + rootPrefixStr + ". " +
				"Insufficient capacity to allocate a new prefix of size: " + strconv.Itoa(desiredSize.(int)) +
				" and with desiredValue of: " + desiredValue.(string) + "\n" +
				"Currently allocated addresses: " + addressesToStr(ipv4prefix.currentResources))
		}
	}

	// iterate over allocated subnets and see if a desired new subnet can be squeezed in
	for _, allocatedSubnet := range currentResourcesStruct {
		allocatedSubnetNum, _ := InetAton(allocatedSubnet.address)
		chunkCapacity := allocatedSubnetNum - possibleSubnetNum

		if desiredValue != nil {
			desiredValueNum, _ := InetAton(desiredValue.(string))
			desiredBroadcastNum, _ := InetAton(inetNtoa(subnetLastAddress(desiredValueNum, newSubnetMask)))

			if chunkCapacity >= newSubnetCapacity && desiredBroadcastNum < allocatedSubnetNum {
				// there is chunk with sufficient capacity between possibleSubnetNum and allocatedSubnet.address
				result["address"] = inetNtoa(desiredValueNum)
				result["prefix"] = newSubnetMask
				result["subnet"] = isSubnet

				return result, nil
			}
		} else {
			if chunkCapacity >= newSubnetCapacity {
				// there is chunk with sufficient capacity between possibleSubnetNum and allocatedSubnet.address
				result["address"] = inetNtoa(possibleSubnetNum)
				result["prefix"] = newSubnetMask
				result["subnet"] = isSubnet

				return result, nil
			}
		}

		// move possible subnet start to a valid address outside allocatedSubnet's addresses and continue the search
		possibleSubnetNum = findNextFreeSubnetAddress(allocatedSubnet, newSubnetMask)
	}

	// check if there is any space left at the end of parent range
	if desiredValue == nil && possibleSubnetNum+newSubnetCapacity <= rootAddressNum+rootCapacity {
		// there sure is some space, use it !
		result["address"] = inetNtoa(possibleSubnetNum)
		result["prefix"] = newSubnetMask
		result["subnet"] = isSubnet
		return result, nil
	}

	var desiredSizeStr string
	if isSubnet.(bool) {
		desiredSizeStr = strconv.Itoa(desiredSize.(int) - 2)
	} else {
		desiredSizeStr = strconv.Itoa(desiredSize.(int))
	}

	return nil, errors.New("Unable to allocate Ipv4 prefix from: " + rootPrefixStr + ". " +
		"Insufficient capacity to allocate a new prefix of size: " + desiredSizeStr + "\n" +
		"Currently allocated addresses: " + addressesToStr(ipv4prefix.currentResources))
}
